### DISCUSSION TITLE
I made a kitten to go to my 'notes' tab and back

### DISCUSSION THREAD

**Discussion by tg0h** [kovidgoyal/kitty#5396] [Show and tell]
Hello, just wanted to share a kitten I created.

I created a kitten to quickly go to a tab with my notes, then get back to whatever tab I was in before.

### Background

I stopped using tmux but wanted a tmux-sessionizer like experience where I had a session which I could navigate in and out of. In this case, instead of sessions, I wanted a tab I could go to and get out of.

### Features

When the kitten is called,
- if a tab named 'wiki' does not exist, create this tab (which opens in my wiki folder)
- if a tab named 'wiki already exists
  - if the current tab is not the 'wiki' tab
    -  focus the wiki tab
  - if the current tab is the 'wiki' tab
    -  if the wiki tab is not the rightmost tab, move it to be the rightmost tab - I do this as I want to display the wiki tab on the right side of my custom [tab bar](https://github.com/kovidgoyal/kitty/discussions/4447#discussioncomment-3470956)
    -  go to the previous tab (doesn't work if the wiki tab has just been moved to the rightmost tab)

### In future
- Display these 'persistent' tabs on the right side of my tab bar. - [DONE](https://github.com/kovidgoyal/kitty/discussions/4447#discussioncomment-3470956)
- Make these tabs harder to close by using separate keymaps to close 'normal' vs 'persistent' tabs - DONE - either use custom kitten or remote control [command](https://github.com/kovidgoyal/kitty/discussions/5406)

### How I made this

This kitten is quite simple — it only needs to get a list of tabs, determine if a tab with a title exists, and then create or go to that tab. (and colour the tab) 

I have never made a kitten before — I spent 7 hours looking around the kitty source code, github issues and discussions, the source code of other kittens, and also spent 12 hours speed watching a udemy python course lol.

To get a list of tabs, I looked at the boss class and tried to see if there was a list of tabs I could loop though to inspect. After 3 hours, I eventually figured out I could use the convenient match_tabs method `boss.match_tabs('title:^wiki')`
 
The kitty remote control commands are very well [documented](https://sw.kovidgoyal.net/kitty/remote-control/#kitty), but I found that the internal apis available to a kitty are not as well documented. I eventually figured out that I could look at the remote control commands in the kitty [repo](https://github.com/kovidgoyal/kitty/tree/master/kitty/rc) to grok what to do (In addition to all the other searching I did). For example, to colour the tab orange, I found that `kitty @ set-tab-color inactive_bg=orange inactive_fg=black` did the trick. With this, I looked at `rc/set_tab_color.py` for how to do this via a kitten.

A random list of references that I used
- https://github.com/kovidgoyal/kitty/discussions/4363
- https://github.com/kovidgoyal/kitty/discussions/3183
- https://github.com/knubie/vim-kitty-navigator

## Code
```python
from typing import List
from kitty.boss import Boss
from kittens.tui.loop import debug
from kittens.tui.handler import result_handler
from kitty.rgb import to_color

def main(args: List[str]):
    pass

def redraw_tab_bar(boss):
    tm = boss.active_tab_manager
    if tm is not None:
        tm.mark_tab_bar_dirty()   

def get_wiki_tab_position(tabs):
    i=1
    for tab in tabs:
        if getattr(tab,"title",None) == 'wiki':
            return i
        i+=1
    return None

def move_wiki_tab_to_end(tabs, boss):
    # assumes that wikiTab is currently focussed
    tabLength = len(tabs)
    wikiTabPosition = get_wiki_tab_position(tabs)
    if wikiTabPosition == tabLength: # wiki tab is already at the rightmost tab
        return
    if (wikiTabPosition and tabLength > wikiTabPosition):
            current_position=wikiTabPosition
            while (current_position < tabLength):
                boss.move_tab_forward() # this moves the currently focussed tab (which may not be wiki )
                current_position+=1

    
@result_handler(no_ui=True)
def handle_result(args: List[str], answer: str, target_window_id: int, boss: Boss) -> None:
    # go to tab named 'wiki'
    # if tab does not exist, create the tab
    # if currently focussed tab is the 'wiki' tab, then go to previous tab
    # change the 'wiki' tab colour

    # print(f'active tab is {boss.active_tab}')
    tm = boss.active_tab_manager
    # print(f'there are {len(tm.tabs)} tabs')
    # for tab in tm.tabs:
    #     print(f'tab id is {getattr(tab,"id","oops")}')

    colourOrange = int(to_color('orange', validate=True))
    colourBlack = int(to_color('black', validate=True))

    is_wiki_focussed=False
    if boss.active_tab.title == 'wiki':
        is_wiki_focussed=True

    tabs = boss.match_tabs('title:^wiki')
    tab = next(tabs, None) # default value for generator to None instead of throwing StopIteration error

    if tab:
        # tab.inactive_bg = colourOrange
        tab.inactive_fg = colourOrange
        tab.active_fg = colourBlack
        tab.active_bg = colourOrange
        redraw_tab_bar(boss)
        if not is_wiki_focussed:
            boss.set_active_tab(tab)
        else:
            # assumes that tab is wiki tab and is currently focussed
            move_wiki_tab_to_end(tm.tabs, boss)
            print('going to prev active tab')
            # boss._move_tab_to(tab, target_window_id)
            # to go to previous active tab
            # unlike goto_tab -1, boss.goto_tab uses 0 instead
            boss.goto_tab(0) # go to previous active tab
    else:
        boss.launch(
        "--type=tab",
        "--title=wiki",
        "--cwd=/Users/tim/src/me/wiki",
        )
```

**Comment by kovidgoyal** [comment]
This should help:
https://github.com/kovidgoyal/kitty/commit/2efb8f6dc29c93ba51fa5add1b1705d9f3fc5a30

**Comment by megalithic** [comment]
This is brilliant! Loving these types of kittens!
